### ISSUE TITLE
crash on upstream connection from envoy to envoy using HTTP1/TLS

### ISSUE THREAD

**Issue by rgs1** [envoyproxy/envoy#6947]
We have the following mesh setup:

```
envoy-A --> H1 over TLS --> envoy-B
```

and today most of the  envoys in cluster`envoy-A` crashed when the `envoy-B`
cluster was hot-restarted.

The stack trace was:

```
Caught Segmentation fault, suspect faulting address 0x28
Backtrace (use tools/stack_decode.py to get line numbers):
#0: __restore_rt [0x7fea8cb4f390] ??:0
#1: Envoy::Http::StreamDecoderWrapper::decodeData() [0xa0791e] bazel-out/k8-opt/bin/external/envoy/source/common/http/_virtual_includes/codec_wrappers_lib/common/http/codec_wrappers.h:35
#2: Envoy::Http::Http1::ClientConnectionImpl::onMessageComplete() [0xad049a] external/envoy/source/common/http/http1/codec_impl.cc:762
#3: Envoy::Http::Http1::ConnectionImpl::onMessageCompleteBase() [0xad012c] external/envoy/source/common/http/http1/codec_impl.cc:476
#4: Envoy::Http::Http1::ConnectionImpl::{lambda()#7}::_FUN() [0xad021d] external/envoy/source/common/http/http1/codec_impl.cc:312
#5: http_parser_execute [0xc5d070] external/com_github_nodejs_http_parser/http_parser.c:1921 (discriminator 1)
#6: Envoy::Http::Http1::ConnectionImpl::dispatchSlice() [0xacdca1] external/envoy/source/common/http/http1/codec_impl.cc:397
#7: Envoy::Http::Http1::ConnectionImpl::dispatch() [0xacf985] external/envoy/source/common/http/http1/codec_impl.cc:381 (discriminator 2)
#8: Envoy::Http::CodecClient::onData() [0xa593f1] external/envoy/source/common/http/codec_client.cc:135
#9: Envoy::Http::CodecClient::CodecReadFilter::onData() [0xa595cd] bazel-out/k8-opt/bin/external/envoy/source/common/http/_virtual_includes/codec_client_lib/common/http/codec_client.h:169
#10: Envoy::Network::FilterManagerImpl::onContinueReading() [0x8e2faa] external/envoy/source/common/network/filter_manager_impl.cc:60
#11: Envoy::Network::ConnectionImpl::onReadReady() [0x8df18a] external/envoy/source/common/network/connection_impl.cc:272
#12: Envoy::Network::ConnectionImpl::onFileEvent() [0x8df94a] external/envoy/source/common/network/connection_impl.cc:492
#13: Envoy::Event::FileEventImpl::assignEvents()::{lambda()#1}::_FUN() [0x8d948a] external/envoy/source/common/event/file_event_impl.cc:66
#14: event_process_active_single_queue.isra.31 [0xc49529] /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/external/com_github_libevent_libevent/event.c:1638
#15: event_base_loop [0xc4b23f] /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/external/com_github_libevent_libevent/event.c:1805
#16: Envoy::Event::DispatcherImpl::run() [0x8d80ba] external/envoy/source/common/event/dispatcher_impl.cc:179
#17: Envoy::Server::InstanceImpl::run() [0x89f934] external/envoy/source/server/server.cc:484
#18: Envoy::MainCommonBase::run() [0x4b5d14] external/envoy/source/exe/main_common.cc:102
#19: main [0x420fb3] external/envoy/source/exe/main.cc:39
#20: [0x7fea8c48b830]
```

We are suspicious of #6578, given it touched lines close to the crash location:

https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/codec_impl.cc#L762

I'll try to get a repro tomorrow.

cc: @fishcakez @derekargueta @sc0ttbeardsley @mattklein123

**Comment by mattklein123** [comment]
@rgs1 almost definitely this. I spent most of my day today figuring this out. :) https://github.com/envoyproxy/envoy/pull/6942

**Comment by rgs1** [comment]
@mattklein123 we should coordinate our debugging parties... thanks for the fix!

**Comment by dio** [comment]
Closed by #6942. Please re-open if otherwise. Thank you!
